### PR TITLE
 flatpak: Mark Endless apps with popular kudo 

### PR DESCRIPTION
**Pull request by uajain**
https://phabricator.endlessm.com/T26507

### CODE REVIEW

**Comment by uajain** [comment]
This basically worked for me.
At some point, I saw some `xb_builder_node_get_child: assertion 'XB_IS_BUILDER_NODE (self)' failed` failures in `--verbose` log, so will be looking into that tomorrow.

File: plugins/core/gs-appstream.c [hunk 1]
```diff
@@ -1408,6 +1402,25 @@ gs_appstream_component_add_keyword (XbBuilderNode *component, const gchar *str)
...
+	g_autoptr(XbBuilderNode) kudos = NULL;
+	g_autoptr(XbBuilderNode) kudo = NULL;
+
+	/* create <kudos> if it does not already exist */
+	kudos = xb_builder_node_get_child (component, "kudos", NULL);
+	if (kudos == NULL)
+		kudos = xb_builder_node_insert (component, "kudos", NULL);
+
+	/* create <kudo>str<kudo> if it does not already exist */
+	kudo = xb_builder_node_get_child (kudo, "kudo", str);
```
Comment (uajain) [inline plugins/core/gs-appstream.c]: This was causing the warnings. Fixed up locally. 
```suggestion
	kudo = xb_builder_node_get_child (kudos, "kudo", str);
```

File: plugins/core/gs-appstream.h [hunk 2]
```diff
@@ -71,5 +71,7 @@ void		 gs_appstream_component_add_icon	(XbBuilderNode	*component,
 							 const gchar	*str);
 void		 gs_appstream_component_add_provide	(XbBuilderNode	*component,
 							 const gchar	*str);
+void		 gs_appstream_component_add_kudo	(XbBuilderNode	*component,
```
Comment (pwithnall) [inline plugins/core/gs-appstream.h]: Shouldn’t this new function be called from somewhere?

Comment (uajain) [reply plugins/core/gs-appstream.h]: Sorry, I made a `fixup/squash/rebase` fiasco and had hunks not pushed on the PR. 
Pushed now.

File: plugins/flatpak/gs-flatpak.c [hunk 3]
```diff
@@ -323,6 +323,22 @@ gs_flatpak_add_flatpak_keyword_cb (XbBuilderFixup *self,
...
 }
 
+static gboolean
+gs_flatpak_mark_endlessm_apps_popular_cb (XbBuilderFixup *self,
+					   XbBuilderNode *bn,
+					   gpointer user_data,
+					   GError **error)
+{
+	if (g_strcmp0 (xb_builder_node_get_element (bn), "component") == 0) {
+		g_autoptr(XbBuilderNode) id = xb_builder_node_get_child (bn, "id", NULL);
```
Comment (pwithnall) [inline plugins/flatpak/gs-flatpak.c]: It’s possible for `id` to be `NULL` if the XML is malformed (a `<component>` doesn’t contain an `<id>`), which will cause a crash.

File: plugins/flatpak/gs-flatpak.c [hunk 4]
```diff
@@ -323,6 +323,22 @@ gs_flatpak_add_flatpak_keyword_cb (XbBuilderFixup *self,
...
+		if (g_str_has_prefix (xb_builder_node_get_text (id), "com.endlessm.") ||
+		    g_str_has_prefix (xb_builder_node_get_text (id), "com.endlessnetwork.")) {
```
Comment (pwithnall) [inline plugins/flatpak/gs-flatpak.c]: It’s possible for `xb_builder_node_get_text()` to return `NULL` (if the XML contains `<id/>` or `<id></id>`), which `g_str_has_prefix()` won’t be happy with.

**Comment by pwithnall** [comment]
[Not needed for now](https://phabricator.endlessm.com/T26507#740922), so I’ll close this.

**Pull request closed without merging**